codecombat/codecombat
# Lifetime migrated to a yearly premium pricing

# Description

We need to migrate from a once off lifetime payment to a yearly subscription option.

## Type of change

On the frontend I am additively only changing the lifetime option to use the subscription flow. The client sends the stripe `planID` to the server to apply any coupons and subscribe the individual.

Thus the majority of the change is passing up a different planID for the yearly subscription.

## Testing process

### Manual testing

Manual testing is done with the stripe test api and test products. Requires setting up local database.

A lot of manual testing has been done with test products in order to make sure this works. Once the schemas are deployed to production I will be able to test this in our test environment.

Cases tested:

 - Regression testing the basic subscription flow.
 - Subscribing with a test product with a yearly subscription.
 - Unsubscribe page correctly unsubscribes and resubscribes without double charging.
 - Unsubscribe page shows the correct recurring payment and end of subscription (matching Stripe dashboard).
 - Server rejects invalid plans.
 - Tested valid and invalid coupons. Invalid coupons cause an internal error and won't happen if database is set up correctly.
 - Tested subscription view with coupon discount from stripe.


### Unit testing

 - [x] Repair unit tests to support the new subscription button. (2 tests broken)

### TODOS

 - [x] Fix the Parent page product load
 - [x] Update tests with the new yearly subscribe button - They use same subscription logic as monthly.
 - [x] Add coupons to the database to reflect what coupons we have in stripe.
 - [x] Make sure coupon discounts are accurately reflected on our `/account/subscription` page


## Screenshots and gifs

![9e3d6fd15d153485f4090d3efbdb8366](https://user-images.githubusercontent.com/15080861/98057527-57acc480-1df7-11eb-9077-e97c8e29f6ca.gif)

![image](https://user-images.githubusercontent.com/15080861/98069125-e7ac3780-1e12-11eb-86ff-612c1ce32487.png)

### Subscription view showing yearly discount

(Test yearly subscription is $2 - thus after discount we get $0.80)

![Screen Shot 2020-11-04 at 11 23 21 AM](https://user-images.githubusercontent.com/15080861/98159016-b6724c80-1e90-11eb-840d-3178273b56b2.png)
